src/assign-graders/LoadBalancer.ts

```diff
@@ -0,0 +1,220 @@
+import { green } from "colors";
+import { Ora } from "ora";
+import { Page } from "puppeteer";
+import { MAIN_URL } from "../common/constants";
+import UserError from "../common/UserError";
+import { Application, Grader } from "./types";
+
+/**
+ * Interact with Greenhouse UI and assign
+ * graders to written inteviews
+ */
+export default class LoadBalancer {
+    private page: Page;
+    private graders: Grader[];
+    private jobs: string[];
+    private spinner: Ora;
+
+    constructor(
+        page: Page,
+        graders: Grader[],
+        selectedJobs: string[],
+        spinner: Ora
+    ) {
+        this.page = page;
+        this.graders = graders;
+        this.jobs = selectedJobs;
+        this.spinner = spinner;
+    }
+
+    /**
+     * Return applications in current page
+     */
+    private async getApplicationsPage() {
+        return await this.page.$$eval(".person", (people) =>
+            people.map((p) => {
+                const applicationID = p.getAttribute("application");
+                const toggleText = p.querySelector(
+                    "a.toggle-interviews"
+                )?.textContent;
+                const toGrade = toggleText?.includes("Scorecard due");
+                const job = p
+                    .querySelector(".job")
+                    // Delete requisition ID next to job name
+                    ?.textContent?.replace(/\(\d+\)/, "")
+                    .trim();
+                const candidate = p.querySelector(".name a")?.textContent;
+
+                if (
+                    applicationID != null &&
+                    candidate != null &&
+                    job != null &&
+                    toGrade != null
+                ) {
+                    return {
+                        applicationID,
+                        candidate,
+                        job,
+                        toGrade,
+                    };
+                }
+            })
+        );
+    }
+
+    /**
+     * Get applications to process one at a time
+     */
+    private async *getApplicationsToProcess() {
```

Comment:
The idea was: the generator abstracts the logic for going through all the pages and it returns an application to process at a time, so the main function only needs to click around. I'm happy to delete it (as I agree it's a bad idea most of the time), but I don't know if I got your idea. Do you mean moving the loop to the `execute` method?

Reply:
Sorry what I mean is simplifying this function, so something like this could work:
```ts
    private async getApplicationsToProcess():Promise<Application[]> {
        const applications = [];
        let nextPageBtn;
        do {
            await this.page.waitForSelector(".person");
            const applicationsPage = await this.getApplicationsPage();

            for (const application of applicationsPage) {
                if (
                    application &&
                    application?.toGrade &&
                    this.jobs.includes(application.job)
                ) {
                    applications.push(application);
                }
            }

            // Keep doing this until there are no more pages
            nextPageBtn = await this.page.$("a.next_page:not(.disabled)");

            await this.page.waitForNavigation(),
            await nextPageBtn.click(),
        } while(nextPageBtn)
        return applications;
    }

Reply:
The problem I see is that the applications returned (because of how I designed this) only contain identifiers (such as id, candidate name...). The responsibility of the main function is to receive the application, select the elements in the DOM and perform actions.

Your code will go through all available pages and return all the applications in all pages. But then when you want to loop over it and assign graders, you'll need to go through all pages again, otherwise it won't find the selectors in the DOM.

The advantage of the generator is that it pauses and returns the control to the caller. So if you're on page 2, for example, it yields the application and we process it, and then it goes back again to get the next application.

Reply:
(I like the `do-while` better than the loop I wrote, btw)

If you find it more readable, what I can do is go with a more classic approach: a loop in the same function. 

- Get applications page
- Assign graders to each of them
- Go to the next page
- Repeat

I went with the generator because it seemed to me like a nice way to hide the "loop over pages" part.

Reply:
Yes, I think going the non generator functions is the wright approach as the code flow is more readable for people who aren't familiar with generator function (like me :))